deps/happy-eyeballs/happy-eyeballs.c

```diff
@@ -0,0 +1,528 @@
+		if (os_event_try(context->race_completed_event) == EAGAIN) {
+			// We are the winner.
+			context->socket_fd = args->sockfd;
+			memcpy(&context->winner_addr, args->address->ai_addr,
+			       sizeof(struct addrinfo));
+			os_event_signal(context->race_completed_event);
+		}
+		pthread_mutex_unlock(&context->winner_mutex);
+		free(args);
+		return NULL;
+	}
+
+failure:
+	// failure, note down the error
+	args->candidate->error = GetSockError();
+	pthread_mutex_lock(&context->winner_mutex);
+	os_event_signal(args->candidate->socket_completed_event);
+	// next, check if we are the last worker running. If so, we'll set the error status and signal the completion
+	// event.
+	bool active = false;
+	for (int i = 0; i < HAPPY_EYEBALLS_MAX_ATTEMPTS && !active; i++) {
+		active = context->candidates[i].valid &&
+			 os_event_try(context->candidates[i]
+					      .socket_completed_event) ==
+				 EAGAIN;
+	}
+	pthread_mutex_unlock(&context->winner_mutex);
+	if (active) {
+		// we're not last...
+		free(args);
+		return NULL;
+	}
+	// Ok, we are last. Coalesce errors and signal completion.
+	context->error_message = "Failed to connect";
+	os_event_signal(context->race_completed_event);
+	free(args);
+	return NULL;
+}
+
+static int launch_worker(struct happy_eyeballs_ctx *context, int index,
+			 struct addrinfo *addr)
+{
+#ifdef _WIN32
+	SOCKET fd = WSASocket(addr->ai_family, SOCK_STREAM, IPPROTO_TCP, NULL,
+			      0, WSA_FLAG_OVERLAPPED);
+#else
+	SOCKET fd = socket(addr->ai_family, SOCK_STREAM, IPPROTO_TCP);
+#endif
+	struct happy_eyeballs_candidate *candidate =
+		&context->candidates[index];
+	candidate->sockfd = fd;
+	struct happy_connect_woker_args *args =
+		(struct happy_connect_woker_args *)malloc(
+			sizeof(struct happy_connect_woker_args));
+	if (args == NULL) {
+		context->error = ENOMEM;
+		context->error_message =
+			"happy-eyeballs: Failed to allocate memory for worker context";
+		return STATUS_FAILURE;
+	}
+	args->sockfd = (int)fd;
+	args->address = addr;
+	args->context = context;
+	args->candidate = candidate;
+
+	int result = os_event_init(&candidate->socket_completed_event,
+				   OS_EVENT_TYPE_MANUAL);
+	if (result != 0) {
+		// failure to create the
+		context->error = result;
+		context->error_message =
+			"happy-eyeballs: Failed to initialize socket completed event";
+		return STATUS_FAILURE;
+	}
+
+	result = pthread_create(&candidate->thread, NULL, happy_connect_worker,
+				args);
+	if (result != 0) {
+		// failure to start the worker thread
+		context->error = result;
+		context->error_message = strerror(result);
+		os_event_destroy(candidate->socket_completed_event);
+		return STATUS_FAILURE;
+	}
+
+	candidate->valid = true;
+	return STATUS_SUCCESS;
+}
+
+int happy_eyeballs_init(struct happy_eyeballs_ctx **context)
+{
+	if (context == NULL) {
+		return STATUS_INVALID_ARGUMENT;
+	}
+
+	*context = (struct happy_eyeballs_ctx *)malloc(
+		sizeof(struct happy_eyeballs_ctx));
+	if (*context == NULL) {
```

Comment:
curlies

Reply:
It's a bit weird that we return an error in two ways. One is FALSE and context NULL. The other is FALSE and context non-null, but filled with error values.

I wonder if it makes sense to have a
```cpp
struct happy_eyeballs_context_error {
	/**
	 * error will be non-zero in case of an error during operation.
	 */
	int error;

	/**
	 * error_message may be set when there is a non-zero error code.
	 */
	char *error_message;
}
```

Then you can have happy_eyeballs_context_error inside the context and happy_eyeballs_context_error  which is just an out param in the _create()

```cpp
int happy_eyeballs_init(struct happy_eyeballs_ctx **context, struct happy_eyeballs_context_error **error)
```

Food for thought

Reply:
Yeah I think this is a good suggestion, will adopt.